docs: data square layout specs update part 1 (#1905)
celestiaorg/celestia-app: specs/src/specs/data_square_layout.md

    1. These transactions contain protobuf encoded [`sdk.Msg`](https://github.com/celestiaorg/cosmos-sdk/blob/v1.14.0-sdk-v0.46.11/types/tx_msg.go#L14-L26)s, which get executed atomically (if one fails they all fail) to update the Celestia state. The complete list of modules, which define the `sdk.Msg`s that the state machine is capable of handling, can be found in the [state machine modules spec](../specs/state_machine_modules.md). Examples include standard cosmos-sdk module messages such as [MsgSend](https://github.com/cosmos/cosmos-sdk/blob/f71df80e93bffbf7ce5fbd519c6154a2ee9f991b/proto/cosmos/bank/v1beta1/tx.proto#L21-L32)), and celestia specific module messages such as [`MsgPayForBlobs`](https://github.com/celestiaorg/celestia-app/blob/v1.0.0-rc2/proto/celestia/blob/v1/tx.proto#L16-L31)
1. Blobs: binary large objects which do not modify the Celestia state, but which are intended for a Celestia application identified with a provided namespace.

> We want to arrange this data into a `k * k` matrix of fixed-sized [shares](../specs/shares.md), which will later be committed to in [Namespace Merkle Trees (NMTs)](https://github.com/celestiaorg/nmt/blob/v0.16.0/docs/spec/nmt.md) so that individual shares in this matrix can be proven to belong to a single data root.

Reviewer: [no change needed][question] for consistency, should we refer to it as "Namepsace" or "Namespaced"  Merkle Trees? 

https://github.com/celestiaorg/nmt README says "Namespaced" but TBH I prefer "Namespace"

Author: I'm good with either, but the [LL paper](https://arxiv.org/pdf/1905.09274.pdf) uses "namespaced"

This, however, requires the block producer to interact with the transaction sender to provide them the starting location of their blob, so that the sender can sign over the commitment based on that starting location. This can be done selectively, but is not ideal as a default for e.g. end-user wallets.

> ### Non-Interaction Rules

Reviewer 2: ```suggestion
### Share Commitment Rules
```
or Blob Commitment Rules

Reviewer: or Blob Share Commitment Rules 🙂

Reviewer 2: Yeah I'm a fan of that

Author: cc also curious what @adlerjohn thinks about the name

we've changed them to "Blob Share Commitment Rules" for now. Personally, I'd opt for something like "Blob Layout Rules" or just "Blob Commitment Rules", but bscr are suitable as well

Author: I'm not too picky though, whatever people can get behind I'm good with

> ### Non-Interaction Rules

Reviewer 2: ```suggestion
### Blob Share Commitment Rules
```

> We impose one additional rule on blob placement to make the possible starting locations of blobs sufficiently predictable and constrained such that users can deterministically compute subtree roots without the need to interact with the block proposer:

Reviewer 2: ```suggestion
To make the possible starting locations of blobs sufficiently predictable and constrained such that users can deterministically compute subtree roots, needed for the `ShareCommitment`s within a PFB, without the need to interact with the block proposer, we impose one additional rule:
```

> > Blobs must start at an index that is a multiple of the `SubtreeWidth`. The `SubtreeWidth` is the length of the blob in shares, divided by the [`SubtreeRootThreshold`](https://github.com/celestiaorg/celestia-app/blob/v1.0.0-rc2/pkg/appconsts/v1/app_consts.go#L6) ([implementation here](https://github.com/celestiaorg/celestia-app/blob/v1.0.0-rc2/pkg/shares/non_interactive_defaults.go#L94-L116)).

Reviewer 2: Isn't `SubtreeWidth` the length of the blob divided by the `SubtreeRootThreshold` and then to the power of two. If we don't want the power of two part then really it's the `SubtreeDepth` (or height lol depending on who you ask).

Also as an additional constraint, a blob may start at the start of the next row which may not necessarily be a multiple of the SubtreeWidth or if SubTreeWidth is too large we use the `BlobMinSquareSize()`

Author: good point, and yeah I agree. I rushed this through last night to at least get an up to date version since other teams are already building on it and need that, but was thinking the same thing. Visuals would go a long way here. https://github.com/celestiaorg/celestia-app/issues/1913

5fed294d871f85f0839540859c4cec56acc3dd1a

> The `SubtreeRootThreshold` is an arbitrary versioned protocol constant that aims to put a soft limit the number of subtree roots included in a blob inclusion proof, as described in [ADR013](../../../docs/architecture/adr-013-non-interactive-default-rules-for-zero-padding.md).

Reviewer 2: ```suggestion
The `SubtreeRootThreshold` is an arbitrary versioned protocol constant that aims to put a soft limit the number of subtree roots included in a blob inclusion proof, as described in [ADR013](../../../docs/architecture/adr-013-non-interactive-default-rules-for-zero-padding.md). A higher SubtreeRoot Threshold means less padding and more tightly packed squares but also means greater proof sizes.
```

> In the constraint mentioned above, the number of rows/columns in the minimum square size should be a power of 2.

Reviewer: [question] is this line still applicable? There is no reference to minimum square size above

Author: resolved in https://github.com/celestiaorg/celestia-app/pull/1905/commits/bfa854b75d11c8ebcd51073e9f03c48798e5b758

> With the above constraint, we can compute subtree roots deterministically. For example, a blob of 128 shares and SRT = 64, must start on a share index that is a multiple of 2 because 128/64 = 2. In this case, there will be a maximum of 1 share of padding between blobs (more on padding below). The maximum subtree width in shares will also be 2, meaning that there will be 2 shares under each subtree root.

Reviewer: [nit] SRT hasn't been used yet so proposal to do

```suggestion
With the above constraint, we can compute subtree roots deterministically. For example, a blob of 128 shares and `SubtreeRootThreshold` (SRT) = 64, must start on a share index that is a multiple of 2 because 128/64 = 2. In this case, there will be a maximum of 1 share of padding between blobs (more on padding below). The maximum subtree width in shares will also be 2, meaning that there will be 2 shares under each subtree root.
```
